### PR TITLE
Add IoU loss

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?
Adding IoU loss

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?

## Who can review?
@LukeWood

### CODE REVIEW

**Comment by quantumalaviya** [comment]
Although a draft PR exists for this (#449), I discussed the API, tests, and the name with @LukeWood, and we decided that there should be a new PR for this.

File: keras_cv/losses/iou_loss.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import tensorflow as tf
+
+from keras_cv.bounding_box import compute_iou
```
Comment (LukeWood) [inline keras_cv/losses/iou_loss.py:18]: prefer:

```
from keras_cv import bounding_box
...
bounding_box.compute_iou()
```

File: keras_cv/losses/iou_loss.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    loss = IoULoss(bounding_box_format = "xyWH")
+    loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+    model.compile(optimizer='adam', loss=keras_cv.losses.IoULoss())
+    ```
+    """
+
+    def __init__(self, bounding_box_format, mode="log", **kwargs):
```
Comment (tanzhenyu) [inline keras_cv/losses/iou_loss.py:58]: Do you anticipate both inputs will be from different formats?

File: keras_cv/losses/iou_loss.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                "IoULoss expects mode to be one of 'linear', 'square' or 'log' "
+                f"Received mode={self.mode}, "
+            )
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+
+        ious = bounding_box.compute_iou(y_true, y_pred, self.bounding_box_format)
+        iou = tf.reduce_mean(ious, axis=[-2, -1])
```
Comment (tanzhenyu) [inline keras_cv/losses/iou_loss.py:74]: This seems to suggest y_pred and y_true have specific shapes, maybe document it here?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/losses/iou_loss.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/losses/iou_loss.py:74]: Even a stage above this: we need to raise error messages if the boxes do not comply to the spec we expect

File: keras_cv/losses/iou_loss.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/losses/iou_loss.py:58]: Yeah, we need to support all bounding box formats.  This is critical to our API consistency, and our continued effort to minimize silent errors.

File: keras_cv/losses/iou_loss.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            - `"squared"`. The loss will be calculated as 1 - iou<sup>2</sup>
+            - `"log"`. The loss will be calculated as -ln(iou)
+            Defaults to "log".
+
+    References:
+        - [UnitBox paper](https://arxiv.org/pdf/1608.01471)
+
+    Sample Usage:
+    ```python
+    y_true = tf.random.uniform([5, 10, 6], minval=0, maxval=10, dtype=tf.dtypes.int32)
```
Comment (LukeWood) [inline keras_cv/losses/iou_loss.py:47]: y_true should not have format [5, 10, 6] -> it should have (5, 10, 5).

no confidence on ground truths.

Also, please use tuples for shapes - thanks!

File: keras_cv/losses/iou_loss.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        self.mode = mode
+
+        if self.mode not in ["linear", "square", "log"]:
+            raise ValueError(
+                "IoULoss expects mode to be one of 'linear', 'square' or 'log' "
+                f"Received mode={self.mode}, "
+            )
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
```
Comment (LukeWood) [inline keras_cv/losses/iou_loss.py:70]: is convert_to_tensor required here?

File: keras_cv/losses/iou_loss.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        y_true = tf.cast(y_true, y_pred.dtype)
```
Comment (LukeWood) [inline keras_cv/losses/iou_loss.py:71]: prefer casting both to self.compute_dtype

Comment (LukeWood) [inline keras_cv/losses/iou_loss.py:71]: lets do the following:

raise value error if y_true.shape does not match [None, None 4+]
same for y_pred

File: keras_cv/losses/iou_loss.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/losses/iou_loss.py:74]: lets replace this with:

```
mean_iou = tf.reduce_mean(ious, axis=[-2, -1])
```

File: keras_cv/losses/iou_loss_test.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        y_pred = [
+            [0, 0, 5, 6, 4, 0.9],
+            [0, 0, 7, 3, 1, 0.76],
+            [4, 5, 5, 6, 4, 0.04],
+            [2, 1, 3, 3, 7, 0.48],
+        ]
+
+        iou_loss = IoULoss(bounding_box_format="xywh")
+
+        self.assertAllClose(iou_loss(y_true, y_pred), 2.0311017)
```
Comment (LukeWood) [inline keras_cv/losses/iou_loss_test.py:78]: how did we come up with this value, can we actually document this?

File: keras_cv/losses/iou_loss_test.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv.losses.iou_loss import IoULoss
+
+
+class IoUTest(tf.test.TestCase):
+    def test_output_shape(self):
```
Comment (LukeWood) [inline keras_cv/losses/iou_loss_test.py:21]: can you add test cases over the reduction strategies, in particular `reduction='none'` is good to test.

**Review by LukeWood — CHANGES_REQUESTED** [review]
Thanks for the PR!  This will be useful in YOLO models.

@quantumalaviya minor comments, but overall LGTM

File: keras_cv/losses/iou_loss.py (hunk 2 above)
Comment (quantumalaviya) [reply keras_cv/losses/iou_loss.py:58]: Does that mean I should add two `bounding_box_format` variables to the constructor, i.e. one for true and one for pred?

Comment (LukeWood) [reply keras_cv/losses/iou_loss.py:58]: No, It’s fine as is feel free to resolve

File: keras_cv/losses/iou_loss.py (hunk 6 above)
Comment (quantumalaviya) [reply keras_cv/losses/iou_loss.py:71]: Raising errors inside the call method? Which method would you recommend for this?

File: keras_cv/losses/iou_loss_test.py (hunk 7 above)
Comment (quantumalaviya) [reply keras_cv/losses/iou_loss_test.py:78]: This was a simple IoU calculation to check the value, should I add a comment for this?

File: keras_cv/losses/iou_loss.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/losses/iou_loss.py:71]: raise ValueError().

if you use`y_true.shape` instead of `tf.shape(y_true)` it will just happen at trace time, not at graph invocation time.  This allows us to perform validation at runtime

File: keras_cv/losses/iou_loss_test.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/losses/iou_loss_test.py:78]: yeah please add a comment just explaining where it came from.

File: keras_cv/losses/iou_loss.py (hunk 6 above)
Comment (quantumalaviya) [reply keras_cv/losses/iou_loss.py:71]: Losses don't seem to have a `self.compute_dtype`. See https://github.com/keras-team/keras/blob/v2.9.0/keras/losses.py#L34-L197. Let me know if you meant something else!

File: keras_cv/losses/iou_loss_test.py (hunk 8 above)
Comment (quantumalaviya) [reply keras_cv/losses/iou_loss_test.py:21]: `test_output_shape_reduction_none` covers this, right?

File: keras_cv/losses/iou_loss.py (hunk 5 above)
Comment (quantumalaviya) [reply keras_cv/losses/iou_loss.py:70]: Yeah, in case we need to support regular lists and numpy matrices? This seems to be the norm throughout most keras loss implementations
